aws_instance: When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore.

### Terraform and AWS Provider Version

```console
Terraform v1.14.6
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v6.35.1
```

### Affected Resource(s) or Data Source(s)

aws_instance

### Expected Behavior

Changing `cpu_options.core_count` or `cpu_options.threads_per_core` should be allowed independently.

### Actual Behavior

Changing either `cpu_options.core_count` or `cpu_options.threads_per_core` alone fails.

### Relevant Error/Panic Output

```console
Error: updating EC2 Instance (<INSTANCE_ID>): modifying CPU options: operation error EC2: ModifyInstanceCpuOptions, https response error StatusCode: 400, RequestID: 73c8ae1b-ec4f-4d77-81a3-1a4a8a9c0f46, api error MissingParameter: When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore.
```


### Sample Terraform Configuration

<details open>
<summary>Click to expand configuration</summary>

```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "6.35.1"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

data "aws_ami" "al2023" {
  most_recent = true
  owners      = ["amazon"]

  filter {
    name   = "name"
    values = ["al2023-ami-2023.*-x86_64"]
  }
}

resource "aws_instance" "example" {
  ami           = data.aws_ami.al2023.id
  instance_type = "c6i.2xlarge"

  cpu_options {
    core_count       = 2
    threads_per_core = 2
  }
}
```
</details>


### Steps to Reproduce

1. Apply configuration
2. Change one of `cpu_options.core_count` or `cpu_options.cpu_options`.
3. Plan, confirming the change of just one of these arguments.
4. Apply configuration

### Debug Logging

<details open>
<summary>Click to expand log output</summary>

```console
2026-03-09T14:43:07.016+0100 [DEBUG] provider.terraform-provider-aws_v6.35.1_x5: HTTP Request Sent: tf_resource_type=aws_instance tf_rpc=ApplyResourceChange http.request.header.authorization="AWS4-HMAC-SHA256 Credential=AKIA************ZMUI/20260309/us-east-1/ec2/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-length;content-type;host;x-amz-date, Signature=*****" http.user_agent="APN/1.0 HashiCorp/1.0 Terraform/1.14.6 (+https://www.terraform.io) terraform-provider-aws/6.35.1 (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go-v2/1.41.3 ua/2.1 os/linux lang/go#1.25.7 md/GOOS#linux md/GOARCH#amd64 api/ec2#1.293.1 m/g" net.peer.name=ec2.us-east-1.amazonaws.com tf_aws.signing_region="" tf_req_id=5513e2a6-dff8-4413-21e6-47113a6b1015 @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.71/logging/tf_logger.go:45 aws.region=us-east-1 http.request.header.amz_sdk_invocation_id=33a002dd-eea8-46a3-aaf8-b1c1907c52fc http.request_content_length=98 http.url=https://ec2.us-east-1.amazonaws.com/ rpc.system.name=aws-api tf_provider_addr=registry.terraform.io/hashicorp/aws
  http.request.body=
  | Action=ModifyInstanceCpuOptions&InstanceId=<INSTANCE_ID>&ThreadsPerCore=2&Version=2016-11-15
   http.request.header.content_type=application/x-www-form-urlencoded http.request.header.x_amz_date=20260309T134307Z rpc.method=ModifyInstanceCpuOptions tf_aws.sdk=aws-sdk-go-v2 tf_mux_provider="*schema.GRPCProviderServer" @module=aws http.method=POST http.request.header.amz_sdk_request="attempt=1; max=25" timestamp="2026-03-09T14:43:07.016+0100"
2026-03-09T14:43:07.928+0100 [DEBUG] provider.terraform-provider-aws_v6.35.1_x5: HTTP Response Received: http.response.header.date="Mon, 09 Mar 2026 13:43:07 GMT" rpc.method=ModifyInstanceCpuOptions rpc.system.name=aws-api tf_aws.sdk=aws-sdk-go-v2 tf_provider_addr=registry.terraform.io/hashicorp/aws @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.71/logging/tf_logger.go:45 aws.region=us-east-1 http.duration=911 http.response.header.vary=accept-encoding tf_aws.signing_region="" tf_mux_provider="*schema.GRPCProviderServer" tf_req_id=5513e2a6-dff8-4413-21e6-47113a6b1015 http.response.header.server=AmazonEC2 http.response.header.strict_transport_security="max-age=31536000; includeSubDomains" tf_rpc=ApplyResourceChange
  http.response.body=
  | <?xml version="1.0" encoding="UTF-8"?>
  | <Response><Errors><Error><Code>MissingParameter</Code><Message>When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore.</Message></Error></Errors><RequestID>7b125bb9-c547-473f-a47b-fb3c5a14baf1</RequestID></Response>
   http.response.header.content_type=text/xml;charset=UTF-8 http.response.header.x_amzn_requestid=7b125bb9-c547-473f-a47b-fb3c5a14baf1 http.status_code=400 tf_resource_type=aws_instance @module=aws http.response.header.cache_control="no-cache, no-store" timestamp="2026-03-09T14:43:07.928+0100"
2026-03-09T14:43:07.929+0100 [DEBUG] provider.terraform-provider-aws_v6.35.1_x5: request failed with unretryable error https response error StatusCode: 400, RequestID: 7b125bb9-c547-473f-a47b-fb3c5a14baf1, api error MissingParameter: When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore.: tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=5513e2a6-dff8-4413-21e6-47113a6b1015 tf_resource_type=aws_instance @caller=github.com/hashicorp/aws-sdk-go-base/v2@v2.0.0-beta.71/logging/tf_logger.go:45 @module=aws rpc.method=ModifyInstanceCpuOptions rpc.system.name=aws-api tf_aws.sdk=aws-sdk-go-v2 tf_rpc=ApplyResourceChange aws.region=us-east-1 tf_mux_provider="*schema.GRPCProviderServer" timestamp="2026-03-09T14:43:07.928+0100"
2026-03-09T14:43:07.930+0100 [ERROR] provider.terraform-provider-aws_v6.35.1_x5: Response contains error diagnostic: @module=sdk.proto diagnostic_summary="updating EC2 Instance (<INSTANCE_ID>): modifying CPU options: operation error EC2: ModifyInstanceCpuOptions, https response error StatusCode: 400, RequestID: 7b125bb9-c547-473f-a47b-fb3c5a14baf1, api error MissingParameter: When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore." tf_proto_version=5.11 tf_resource_type=aws_instance tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.30.0/tfprotov5/internal/diag/diagnostics.go:58 diagnostic_detail="" diagnostic_severity=ERROR tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=5513e2a6-dff8-4413-21e6-47113a6b1015 timestamp="2026-03-09T14:43:07.930+0100"
```
</details>


### GenAI / LLM Assisted Development

n/a

### Important Facts and References

_No response_

### Would you like to implement a fix?

Yes

--- a/.changelog/46879.txt
+++ b/.changelog/46879.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_instance: Fix ` MissingParameter: When specifying CpuOptions you must specify both CoreCount and ThreadsPerCore` errors when updating `cpu_options.core_count` or `cpu_options.threads_per_core`
+```

--- a/internal/service/ec2/ec2_instance.go
+++ b/internal/service/ec2/ec2_instance.go
@@ -1993,16 +1993,17 @@ func resourceInstanceUpdate(ctx context.Context, d *schema.ResourceData, meta an
 				InstanceId: aws.String(d.Id()),
 			}
 
-			if d.HasChange("cpu_options.0.core_count") {
+			// The AWS API requires both CoreCount and ThreadsPerCore to be specified
+			// together when modifying CPU options, so always send both when either changes.
+			// Both attributes are Optional+Computed, so even if only one is in the
+			// configuration, the other will be populated from state (read back from AWS).
+			if d.HasChange("cpu_options.0.core_count") || d.HasChange("cpu_options.0.threads_per_core") {
 				input.CoreCount = aws.Int32(int32(tfMap["core_count"].(int)))
+				input.ThreadsPerCore = aws.Int32(int32(tfMap["threads_per_core"].(int)))
 			}
 
 			if d.HasChange("cpu_options.0.nested_virtualization") {
 				input.NestedVirtualization = awstypes.NestedVirtualizationSpecification(tfMap["nested_virtualization"].(string))
-			}
-
-			if d.HasChange("cpu_options.0.threads_per_core") {
-				input.ThreadsPerCore = aws.Int32(int32(tfMap["threads_per_core"].(int)))
 			}
 
 			_, err := conn.ModifyInstanceCpuOptions(ctx, &input)

--- a/internal/service/ec2/ec2_instance.go
+++ b/internal/service/ec2/ec2_instance.go
@@ -1997,7 +1997,7 @@ func resourceInstanceUpdate(ctx context.Context, d *schema.ResourceData, meta an
 			// together when modifying CPU options, so always send both when either changes.
 			// Both attributes are Optional+Computed, so even if only one is in the
 			// configuration, the other will be populated from state (read back from AWS).
-			if d.HasChange("cpu_options.0.core_count") || d.HasChange("cpu_options.0.threads_per_core") {
+			if d.HasChanges("cpu_options.0.core_count", "cpu_options.0.threads_per_core") {
 				input.CoreCount = aws.Int32(int32(tfMap["core_count"].(int)))
 				input.ThreadsPerCore = aws.Int32(int32(tfMap["threads_per_core"].(int)))
 			}

--- a/internal/service/ec2/ec2_instance_test.go
+++ b/internal/service/ec2/ec2_instance_test.go
@@ -4966,6 +4966,122 @@ func TestAccEC2Instance_cpuOptionsCoreThreads(t *testing.T) {
 	})
 }
 
+func TestAccEC2Instance_cpuOptionsCoreCountOnly(t *testing.T) {
+	ctx := acctest.Context(t)
+	var v awstypes.Instance
+	resourceName := "aws_instance.test"
+	rName := acctest.RandomWithPrefix(t, acctest.ResourcePrefix)
+	var (
+		originalCoreCount      int32 = 2
+		updatedCoreCount       int32 = 3
+		originalThreadsPerCore int32 = 2
+	)
+
+	acctest.ParallelTest(ctx, t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.EC2ServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckInstanceDestroy(ctx, t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccInstanceConfig_cpuOptionsCoreThreads(rName, originalCoreCount, originalThreadsPerCore),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInstanceExists(ctx, t, resourceName, &v),
+				),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
+					},
+				},
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(resourceName, tfjsonpath.New("cpu_options"), knownvalue.ListExact([]knownvalue.Check{knownvalue.ObjectPartial(map[string]knownvalue.Check{
+						"core_count":       knownvalue.Int32Exact(originalCoreCount),
+						"threads_per_core": knownvalue.Int32Exact(originalThreadsPerCore),
+					})})),
+				},
+			},
+			{
+				// Update only core_count, keeping threads_per_core the same.
+				// This verifies that the provider sends both CoreCount and ThreadsPerCore
+				// to the AWS API even when only one of them changes.
+				Config: testAccInstanceConfig_cpuOptionsCoreThreads(rName, updatedCoreCount, originalThreadsPerCore),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInstanceExists(ctx, t, resourceName, &v),
+				),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionUpdate),
+					},
+				},
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(resourceName, tfjsonpath.New("cpu_options"), knownvalue.ListExact([]knownvalue.Check{knownvalue.ObjectPartial(map[string]knownvalue.Check{
+						"core_count":       knownvalue.Int32Exact(updatedCoreCount),
+						"threads_per_core": knownvalue.Int32Exact(originalThreadsPerCore),
+					})})),
+				},
+			},
+		},
+	})
+}
+
+func TestAccEC2Instance_cpuOptionsThreadsPerCoreOnly(t *testing.T) {
+	ctx := acctest.Context(t)
+	var v awstypes.Instance
+	resourceName := "aws_instance.test"
+	rName := acctest.RandomWithPrefix(t, acctest.ResourcePrefix)
+	var (
+		originalCoreCount      int32 = 2
+		originalThreadsPerCore int32 = 2
+		updatedThreadsPerCore  int32 = 1
+	)
+
+	acctest.ParallelTest(ctx, t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.EC2ServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckInstanceDestroy(ctx, t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccInstanceConfig_cpuOptionsCoreThreads(rName, originalCoreCount, originalThreadsPerCore),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInstanceExists(ctx, t, resourceName, &v),
+				),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
+					},
+				},
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(resourceName, tfjsonpath.New("cpu_options"), knownvalue.ListExact([]knownvalue.Check{knownvalue.ObjectPartial(map[string]knownvalue.Check{
+						"core_count":       knownvalue.Int32Exact(originalCoreCount),
+						"threads_per_core": knownvalue.Int32Exact(originalThreadsPerCore),
+					})})),
+				},
+			},
+			{
+				// Update only threads_per_core, keeping core_count the same.
+				// This verifies that the provider sends both CoreCount and ThreadsPerCore
+				// to the AWS API even when only one of them changes.
+				Config: testAccInstanceConfig_cpuOptionsCoreThreads(rName, originalCoreCount, updatedThreadsPerCore),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInstanceExists(ctx, t, resourceName, &v),
+				),
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionUpdate),
+					},
+				},
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownValue(resourceName, tfjsonpath.New("cpu_options"), knownvalue.ListExact([]knownvalue.Check{knownvalue.ObjectPartial(map[string]knownvalue.Check{
+						"core_count":       knownvalue.Int32Exact(originalCoreCount),
+						"threads_per_core": knownvalue.Int32Exact(updatedThreadsPerCore),
+					})})),
+				},
+			},
+		},
+	})
+}
+
 func TestAccEC2Instance_cpuOptionsCoreThreadsUnspecifiedToSpecified(t *testing.T) {
 	ctx := acctest.Context(t)
 	var v awstypes.Instance

